docs: add note about opting into PEP 621 (#1065)
pypa/cibuildwheel: docs/options.md

`setup.py` `setup(python_requires="...")`. If not found, cibuildwheel assumes
the package is compatible with all versions of Python that it can build.

!!! note

>     Rather than using this option, it's recommended you set this value
>     statically in a way that your build backend can use it, too. This ensures
>     that your package's metadata is correct when published on PyPI.

Reviewer: ```suggestion
    Rather than using this option, it's recommended you set this value
    statically in a way that your build backend can use it, too. This ensures
    that your package's metadata is correct when published on PyPI. This option
    is provided by cibuildwheel as an emergency override, and therefore is only
    available in environment variable form.
```

Reviewer 2: ```suggestion
    Rather than using this environment variable, it's recommended you set this value
    statically in a way that your build backend can use it, too. This ensures
    that your package's metadata is correct when published on PyPI. This override
    is provided by cibuildwheel only for emergencies, and therefore is only
    available in environment variable form.
```

I suggest a modest rewrite of the above suggestion to be more clear, avoid repetition and read more cleanly. (In particular, it confused me as to what exactly "this option" was referring to, since it was several paragraphs distant from the original reference, and the intervening ones had talked about the static methods of setting this value)

Reviewer 3: Thanks both. I'll take @CAM-Gerlach's version, but just tweak 'emergencies', as it seems a bit strong. 

```suggestion
    Rather than using this environment variable, it's recommended you set this value
    statically in a way that your build backend can use it, too. This ensures
    that your package's metadata is correct when published on PyPI. This 
    cibuildwheel-specific option is provided as an override, and therefore is only
    available in environment variable form.
```

Reviewer 2: Sounds good; I wasn't too fond of that wording either to be honest.

      - If you have a `pyproject.toml` containing a `[project]` table, you can
        specify `requires-python` there.

        ```toml
        [project]

>         requires-python = ">=3.6"

Reviewer: ```suggestion
        ...
        requires-python = ">=3.6"
```

        ```

>         Note that build backend support for the `[project]` table is still patchy, and
>         adding `[project]` to `pyproject.toml` can change the behaviour of your build
>         (e.g. setuptools may ignore `install_requires` specified via `setup.py` or
>         `setup.cfg`). Make sure to double-check the build after adding.

Reviewer: ```suggestion
        Note that not all build backends fully support using a `[project]` table yet;
        specifically setuptools just added experimental support in version 61.
        Adding `[project]` to `pyproject.toml` requires all the other supported
        values to be specified there, or to be listed in `dynamic`.
```

Reviewer 3: I'll take @henryiii's version, poetry isn't very relevant to wheel building, and it's the existence of the `[project]` table that invokes the change in behaviour.

>       - If you're using setuptools, [you can set this value in `setup.cfg` or `setup.py`](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#python-requirement)
>         and cibuildwheel will read it from there.

Reviewer: ```suggestion
      - If you're using setuptools, [you can set this value in `setup.cfg` (preferred) or `setup.py`](https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#python-requirement)
        and cibuildwheel will read it from there.
```

!!! note
    Rather than using this option, it's recommended you set
    `project.requires-python` in `pyproject.toml` instead:

    ```toml
    [project]
    requires-python = ">=3.6"
    ```

    A few things to note:

      - If your project didn't already have a `pyproject.toml`, be sure to
        set the build system to your build backend.

        ```toml
        [build-system]
        requires = ["setuptools>=42", "wheel"]
        build-backend = "setuptools.build_meta"
        ```

      - If you didn't already have a `pyproject.toml` that had a `[project]`

>         table, you should migrate values from `setup.py` or `setup.cfg`, or

Author: > you should migrate

My only remark is that support for PEP 621 was just added last Friday, so there might be sharp edges left.

---

The following is just maintainer babbling 😅:

We can actively start encouraging people to migrate to it, however in setuptools we decided to keep the `experimental` status for a while to guarantee that we have the freedom to modify how we implement it before committing to an specific design.

Changes in interface will not occur in the `[project]` table (after all it is standardized), but if the users reach out for `[tool.setuptools]` there is no guarantee the fields there will stay unchanged between versions.

If in the following days we are reported that the releases seriously break important packages for the community we might also consider yanking it and backing off a bit while we re-implement any problematic bits.

Reviewer 3: Thanks for reviewing !

> My only remark is that support for PEP 621 was just added last Friday, so there might be sharp edges left.

Hm. Yes okay I see. So what's the conservative route for users with a `[project]` table? Would listing options in `dynamic` be the safest way to go?

Author: If the users specify as dynamic the fields that they are providing via `setup.py`, that would be safe.

I think the parts of PEP 621 important to keep in mind are:

> Build back-ends MUST raise an error if the metadata specifies a field in dynamic but the build back-end was unable to provide the data for it.

> Build back-ends MUST raise an error if the metadata specifies a field statically as well as being listed in dynamic.

> If the metadata does not list a field in dynamic, then a build back-end CANNOT fill in

>         [list them in the the `dynamic` field](https://peps.python.org/pep-0621/#dynamic).
>         The values to include and the format is listed in
>         [PEP 621](https://peps.python.org/pep-0621/#details)

Reviewer 2: ```suggestion
        list them in the [the `dynamic` field](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#dynamic).
        The values to include and the format is listed in the
        [project source metadata specification](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#dynamic)
        (as originally defined in PEP 621).
```

Reviewer 3: thanks!
